File: iOS/SnowTeeth/Utilities/LoopTracker.swift
//
//  LoopTracker.swift
//  SnowTeeth
//
//  Loop detection algorithm for tracking when user completes laps/loops
//

import Foundation
import CoreLocation

struct Loop {
    let id: Int
    let points: [LocationData]
    let simplifiedPoints: [LocationData]
    let distance: Double
    let duration: TimeInterval
    let startTime: Int64
    let endTime: Int64
}

class LoopTracker {
    // Algorithm constants (from shared/algorithms/loop_detection.md)
    private static let loopStartThreshold: Double = 20.0  // meters
    private static let lapThreshold: Double = 25.0         // meters
    private static let minLoopTime: TimeInterval = 30.0    // seconds
    private static let minLoopDistance: Double = 100.0     // meters
    private static let simplificationEpsilon: Double = 10.0 // meters
    private static let lengthTolerance: Double = 0.20       // 20%
    private static let similarityThreshold: Double = 30.0   // meters

    private enum Phase {
        case detectingFirst
        case trackingLaps
    }

    private var phase: Phase = .detectingFirst
    private var allPoints: [LocationData] = []
    private var referenceLoop: Loop?
    private var lapPoint: LocationData?
    private var currentLoopStartIndex: Int = 0
    private var completedLoops: [Loop] = []
    private var spatialIndex: SpatialGrid = SpatialGrid()

    /// Add a new GPS point and check for loop completion
    /// Returns the number of completed loops if a new loop was detected, nil otherwise
    func addPoint(_ point: LocationData) -> Int? {
        allPoints.append(point)
        let currentIndex = allPoints.count - 1

        // Add to spatial index
        spatialIndex.insert(point: point, index: currentIndex)

        switch phase {
        case .detectingFirst:
            return checkForFirstLoop(currentIndex: currentIndex)

        case .trackingLaps:
            return checkForNextLap(currentIndex: currentIndex)
        }
    }

    /// Returns the total number of completed loops
    func getLoopCount() -> Int {
        return completedLoops.count
    }

    // MARK: - Phase 1: First Loop Detection

    private func checkForFirstLoop(currentIndex: Int) -> Int? {
        let currentPoint = allPoints[currentIndex]

        // Don't check points from the last 30 seconds to avoid premature detection
        let minTimeDiff: Int64 = 30000 // 30 seconds in milliseconds

        // Find nearby points that are old enough
        let nearbyIndices = spatialIndex.findNearbyPoints(point: currentPoint, radius: Self.loopStartThreshold)

        for nearbyIndex in nearbyIndices {
            let nearbyPoint = allPoints[nearbyIndex]

            // Check time difference
            if abs(currentPoint.timestamp - nearbyPoint.timestamp) < minTimeDiff {
                continue
            }

            let distance = calculateDistance(point1: currentPoint, point2: nearbyPoint)

            if distance <= Self.loopStartThreshold {
                // Found first loop closure!
                let loopPoints = Array(allPoints[nearbyIndex...currentIndex])

                // Validate loop
                if isValidLoop(points: loopPoints) {
                    // Create reference loop
                    let loop = createLoop(
                        id: 1,
                        points: loopPoints,
                        startIndex: nearbyIndex,
                        endIndex: currentIndex
                    )

                    referenceLoop = loop
                    lapPoint = nearbyPoint
                    currentLoopStartIndex = currentIndex
                    completedLoops.append(loop)
                    phase = .trackingLaps

                    return completedLoops.count
                }
            }
        }

        return nil
    }

    // MARK: - Phase 2: Subsequent Loop Detection

    private func checkForNextLap(currentIndex: Int) -> Int? {
        guard let lapPoint = lapPoint else { return nil }

        let currentPoint = allPoints[currentIndex]
        let distanceToLapPoint = calculateDistance(point1: currentPoint, point2: lapPoint)

        if distanceToLapPoint <= Self.lapThreshold {
            // Potential lap completion
            let candidateLoopPoints = Array(allPoints[currentLoopStartIndex...currentIndex])

            // Validate as a loop
            guard isValidLoop(points: candidateLoopPoints) else {
                return nil
            }

            // Check if it matches the reference loop
            if let referenceLoop = referenceLoop,
               loopsAreSimilar(loop1: candidateLoopPoints, loop2: referenceLoop.points) {
                // Valid matching loop!
                let loop = createLoop(
                    id: completedLoops.count + 1,
                    points: candidateLoopPoints,
                    startIndex: currentLoopStartIndex,
                    endIndex: currentIndex
                )

                completedLoops.append(loop)
                currentLoopStartIndex = currentIndex

                return completedLoops.count
            }
        }

        return nil
    }

    // MARK: - Validation and Comparison

    private func isValidLoop(points: [LocationData]) -> Bool {
        guard points.count >= 2 else { return false }

        // Check minimum time
        let duration = TimeInterval(points.last!.timestamp - points.first!.timestamp) / 1000.0
        if duration < Self.minLoopTime {
            return false
        }

        // Check minimum distance
        let distance = calculateTotalDistance(points: points)
        if distance < Self.minLoopDistance {
            return false
        }

        return true
    }

    private func loopsAreSimilar(loop1: [LocationData], loop2: [LocationData]) -> Bool {
        // Step 1: Simplify both loops
        let simplified1 = douglasPeucker(points: loop1, epsilon: Self.simplificationEpsilon)
        let simplified2 = douglasPeucker(points: loop2, epsilon: Self.simplificationEpsilon)

        // Step 2: Length check
        let length1 = calculateTotalDistance(points: loop1)
        let length2 = calculateTotalDistance(points: loop2)
        let lengthRatio = abs(length1 - length2) / max(length1, length2)

        if lengthRatio > Self.lengthTolerance {
            return false
        }

        // Step 3: Hausdorff-style distance
        let forwardDist = averageDistanceToClosestPoint(from: simplified1, to: simplified2)
        let backwardDist = averageDistanceToClosestPoint(from: simplified2, to: simplified1)
        let maxDeviation = max(forwardDist, backwardDist)

        return maxDeviation < Self.similarityThreshold
    }

    // MARK: - Helper Methods

    private func createLoop(id: Int, points: [LocationData], startIndex: Int, endIndex: Int) -> Loop {
        let simplified = douglasPeucker(points: points, epsilon: Self.simplificationEpsilon)
        let distance = calculateTotalDistance(points: points)
        let duration = TimeInterval(points.last!.timestamp - points.first!.timestamp) / 1000.0

        return Loop(
            id: id,
            points: points,
            simplifiedPoints: simplified,
            distance: distance,
            duration: duration,
            startTime: points.first!.timestamp,
            endTime: points.last!.timestamp
        )
    }

    private func calculateDistance(point1: LocationData, point2: LocationData) -> Double {
        let earthRadiusMeters: Double = 6371000.0

        let lat1Rad = point1.latitude * .pi / 180.0
        let lat2Rad = point2.latitude * .pi / 180.0
        let deltaLat = (point2.latitude - point1.latitude) * .pi / 180.0
        let deltaLon = (point2.longitude - point1.longitude) * .pi / 180.0

        let a = sin(deltaLat / 2) * sin(deltaLat / 2) +
                cos(lat1Rad) * cos(lat2Rad) *
                sin(deltaLon / 2) * sin(deltaLon / 2)

        let c = 2 * atan2(sqrt(a), sqrt(1 - a))

        return earthRadiusMeters * c
    }

    private func calculateTotalDistance(points: [LocationData]) -> Double {
        guard points.count >= 2 else { return 0.0 }

        var total: Double = 0.0
        for i in 1..<points.count {
            total += calculateDistance(point1: points[i-1], point2: points[i])
        }
        return total
    }

    private func averageDistanceToClosestPoint(from: [LocationData], to: [LocationData]) -> Double {
        guard !from.isEmpty && !to.isEmpty else { return 0.0 }

        var totalDistance: Double = 0.0

        for point in from {
            var minDistance = Double.infinity
            for otherPoint in to {
                let distance = calculateDistance(point1: point, point2: otherPoint)
                minDistance = min(minDistance, distance)
            }
            totalDistance += minDistance
        }

        return totalDistance / Double(from.count)
    }

    // Douglas-Peucker algorithm for path simplification
    private func douglasPeucker(points: [LocationData], epsilon: Double) -> [LocationData] {
        guard points.count > 2 else { return points }

        // Find the point with maximum distance
        var maxDistance: Double = 0.0
        var maxIndex: Int = 0
        let end = points.count - 1

        for i in 1..<end {
            let distance = perpendicularDistance(
                point: points[i],
                lineStart: points[0],
                lineEnd: points[end]
            )

            if distance > maxDistance {
                maxDistance = distance
                maxIndex = i
            }
        }

        // If max distance is greater than epsilon, recursively simplify
        if maxDistance > epsilon {
            let left = douglasPeucker(points: Array(points[0...maxIndex]), epsilon: epsilon)
            let right = douglasPeucker(points: Array(points[maxIndex...end]), epsilon: epsilon)

            // Combine results (remove duplicate middle point)
            return left + right.dropFirst()
        } else {
            // All points in between can be removed
            return [points[0], points[end]]
        }
    }

    private func perpendicularDistance(point: LocationData, lineStart: LocationData, lineEnd: LocationData) -> Double {
        // Calculate perpendicular distance from point to line
        let x0 = point.latitude
        let y0 = point.longitude
        let x1 = lineStart.latitude
        let y1 = lineStart.longitude
        let x2 = lineEnd.latitude
        let y2 = lineEnd.longitude

        let numerator = abs((y2 - y1) * x0 - (x2 - x1) * y0 + x2 * y1 - y2 * x1)
        let denominator = sqrt(pow(y2 - y1, 2) + pow(x2 - x1, 2))

        if denominator == 0 {
            return calculateDistance(point1: point, point2: lineStart)
        }

        // Convert to meters (rough approximation)
        let degreesToMeters = 111319.9 // meters per degree at equator
        return (numerator / denominator) * degreesToMeters
    }
}

// MARK: - Spatial Grid for Efficient Proximity Queries

class SpatialGrid {
    private static let cellSize: Double = 50.0 // 50 meters per cell

    private var grid: [GridCell: [Int]] = [:]

    struct GridCell: Hashable {
        let x: Int
        let y: Int
    }

    func insert(point: LocationData, index: Int) {
        let cell = getCell(point: point)
        grid[cell, default: []].append(index)
    }

    func findNearbyPoints(point: LocationData, radius: Double) -> [Int] {
        let centerCell = getCell(point: point)
        var nearbyIndices: [Int] = []

        // Check 3x3 grid of cells
        for dx in -1...1 {
            for dy in -1...1 {
                let cell = GridCell(x: centerCell.x + dx, y: centerCell.y + dy)
                if let indices = grid[cell] {
                    nearbyIndices.append(contentsOf: indices)
                }
            }
        }

        return nearbyIndices
    }

    private func getCell(point: LocationData) -> GridCell {
        // Convert lat/lon to approximate meters (rough calculation)
        let metersPerDegree = 111319.9
        let x = Int(point.latitude * metersPerDegree / Self.cellSize)
        let y = Int(point.longitude * metersPerDegree / Self.cellSize)
        return GridCell(x: x, y: y)
    }
}

File: android/app/src/main/java/com/snowteeth/app/util/LoopTracker.kt
package com.snowteeth.app.util

import com.snowteeth.app.model.LocationData
import kotlin.math.*

data class Loop(
    val id: Int,
    val points: List<LocationData>,
    val simplifiedPoints: List<LocationData>,
    val distance: Double,
    val duration: Long,
    val startTime: Long,
    val endTime: Long
)

class LoopTracker {

    companion object {
        // Algorithm constants (from shared/algorithms/loop_detection.md)
        private const val LOOP_START_THRESHOLD = 20.0  // meters
        private const val LAP_THRESHOLD = 25.0         // meters
        private const val MIN_LOOP_TIME = 30000L       // milliseconds
        private const val MIN_LOOP_DISTANCE = 100.0    // meters
        private const val SIMPLIFICATION_EPSILON = 10.0 // meters
        private const val LENGTH_TOLERANCE = 0.20       // 20%
        private const val SIMILARITY_THRESHOLD = 30.0   // meters
        private const val EARTH_RADIUS_METERS = 6371000.0
    }

    private enum class Phase {
        DETECTING_FIRST,
        TRACKING_LAPS
    }

    private var phase = Phase.DETECTING_FIRST
    private val allPoints = mutableListOf<LocationData>()
    private var referenceLoop: Loop? = null
    private var lapPoint: LocationData? = null
    private var currentLoopStartIndex = 0
    private val completedLoops = mutableListOf<Loop>()
    private val spatialIndex = SpatialGrid()

    /**
     * Add a new GPS point and check for loop completion
     * Returns the number of completed loops if a new loop was detected, null otherwise
     */
    fun addPoint(point: LocationData): Int? {
        allPoints.add(point)
        val currentIndex = allPoints.size - 1

        // Add to spatial index
        spatialIndex.insert(point, currentIndex)

        return when (phase) {
            Phase.DETECTING_FIRST -> checkForFirstLoop(currentIndex)
            Phase.TRACKING_LAPS -> checkForNextLap(currentIndex)
        }
    }

    /**
     * Returns the total number of completed loops
     */
    fun getLoopCount(): Int = completedLoops.size

    // MARK: - Phase 1: First Loop Detection

    private fun checkForFirstLoop(currentIndex: Int): Int? {
        val currentPoint = allPoints[currentIndex]

        // Don't check points from the last 30 seconds to avoid premature detection
        val minTimeDiff = 30000L // 30 seconds in milliseconds

        // Find nearby points that are old enough
        val nearbyIndices = spatialIndex.findNearbyPoints(currentPoint, LOOP_START_THRESHOLD)

        for (nearbyIndex in nearbyIndices) {
            val nearbyPoint = allPoints[nearbyIndex]

            // Check time difference
            if (abs(currentPoint.timestamp - nearbyPoint.timestamp) < minTimeDiff) {
                continue
            }

            val distance = calculateDistance(currentPoint, nearbyPoint)

            if (distance <= LOOP_START_THRESHOLD) {
                // Found first loop closure!
                val loopPoints = allPoints.subList(nearbyIndex, currentIndex + 1).toList()

                // Validate loop
                if (isValidLoop(loopPoints)) {
                    // Create reference loop
                    val loop = createLoop(1, loopPoints, nearbyIndex, currentIndex)

                    referenceLoop = loop
                    lapPoint = nearbyPoint
                    currentLoopStartIndex = currentIndex
                    completedLoops.add(loop)
                    phase = Phase.TRACKING_LAPS

                    return completedLoops.size
                }
            }
        }

        return null
    }

    // MARK: - Phase 2: Subsequent Loop Detection

    private fun checkForNextLap(currentIndex: Int): Int? {
        val lapPoint = this.lapPoint ?: return null

        val currentPoint = allPoints[currentIndex]
        val distanceToLapPoint = calculateDistance(currentPoint, lapPoint)

        if (distanceToLapPoint <= LAP_THRESHOLD) {
            // Potential lap completion
            val candidateLoopPoints = allPoints.subList(currentLoopStartIndex, currentIndex + 1).toList()

            // Validate as a loop
            if (!isValidLoop(candidateLoopPoints)) {
                return null
            }

            // Check if it matches the reference loop
            val referenceLoop = this.referenceLoop
            if (referenceLoop != null && loopsAreSimilar(candidateLoopPoints, referenceLoop.points)) {
                // Valid matching loop!
                val loop = createLoop(
                    completedLoops.size + 1,
                    candidateLoopPoints,
                    currentLoopStartIndex,
                    currentIndex
                )

                completedLoops.add(loop)
                currentLoopStartIndex = currentIndex

                return completedLoops.size
            }
        }

        return null
    }

    // MARK: - Validation and Comparison

    private fun isValidLoop(points: List<LocationData>): Boolean {
        if (points.size < 2) return false

        // Check minimum time
        val duration = points.last().timestamp - points.first().timestamp
        if (duration < MIN_LOOP_TIME) {
            return false
        }

        // Check minimum distance
        val distance = calculateTotalDistance(points)
        if (distance < MIN_LOOP_DISTANCE) {
            return false
        }

        return true
    }

    private fun loopsAreSimilar(loop1: List<LocationData>, loop2: List<LocationData>): Boolean {
        // Step 1: Simplify both loops
        val simplified1 = douglasPeucker(loop1, SIMPLIFICATION_EPSILON)
        val simplified2 = douglasPeucker(loop2, SIMPLIFICATION_EPSILON)

        // Step 2: Length check
        val length1 = calculateTotalDistance(loop1)
        val length2 = calculateTotalDistance(loop2)
        val lengthRatio = abs(length1 - length2) / maxOf(length1, length2)

        if (lengthRatio > LENGTH_TOLERANCE) {
            return false
        }

        // Step 3: Hausdorff-style distance
        val forwardDist = averageDistanceToClosestPoint(simplified1, simplified2)
        val backwardDist = averageDistanceToClosestPoint(simplified2, simplified1)
        val maxDeviation = maxOf(forwardDist, backwardDist)

        return maxDeviation < SIMILARITY_THRESHOLD
    }

    // MARK: - Helper Methods

    private fun createLoop(id: Int, points: List<LocationData>, startIndex: Int, endIndex: Int): Loop {
        val simplified = douglasPeucker(points, SIMPLIFICATION_EPSILON)
        val distance = calculateTotalDistance(points)
        val duration = points.last().timestamp - points.first().timestamp

        return Loop(
            id = id,
            points = points,
            simplifiedPoints = simplified,
            distance = distance,
            duration = duration,
            startTime = points.first().timestamp,
            endTime = points.last().timestamp
        )
    }

    private fun calculateDistance(point1: LocationData, point2: LocationData): Double {
        val lat1Rad = Math.toRadians(point1.latitude)
        val lat2Rad = Math.toRadians(point2.latitude)
        val deltaLat = Math.toRadians(point2.latitude - point1.latitude)
        val deltaLon = Math.toRadians(point2.longitude - point1.longitude)

        val a = sin(deltaLat / 2).pow(2) +
                cos(lat1Rad) * cos(lat2Rad) *
                sin(deltaLon / 2).pow(2)

        val c = 2 * atan2(sqrt(a), sqrt(1 - a))

        return EARTH_RADIUS_METERS * c
    }

    private fun calculateTotalDistance(points: List<LocationData>): Double {
        if (points.size < 2) return 0.0

        var total = 0.0
        for (i in 1 until points.size) {
            total += calculateDistance(points[i - 1], points[i])
        }
        return total
    }

    private fun averageDistanceToClosestPoint(from: List<LocationData>, to: List<LocationData>): Double {
        if (from.isEmpty() || to.isEmpty()) return 0.0

        var totalDistance = 0.0

        for (point in from) {
            var minDistance = Double.MAX_VALUE
            for (otherPoint in to) {
                val distance = calculateDistance(point, otherPoint)
                minDistance = minOf(minDistance, distance)
            }
            totalDistance += minDistance
        }

        return totalDistance / from.size
    }

    // Douglas-Peucker algorithm for path simplification
    private fun douglasPeucker(points: List<LocationData>, epsilon: Double): List<LocationData> {
        if (points.size <= 2) return points

        // Find the point with maximum distance
        var maxDistance = 0.0
        var maxIndex = 0
        val end = points.size - 1

        for (i in 1 until end) {
            val distance = perpendicularDistance(
                points[i],
                points[0],
                points[end]
            )

            if (distance > maxDistance) {
                maxDistance = distance
                maxIndex = i
            }
        }

        // If max distance is greater than epsilon, recursively simplify
        return if (maxDistance > epsilon) {
            val left = douglasPeucker(points.subList(0, maxIndex + 1), epsilon)
            val right = douglasPeucker(points.subList(maxIndex, end + 1), epsilon)

            // Combine results (remove duplicate middle point)
            left + right.drop(1)
        } else {
            // All points in between can be removed
            listOf(points[0], points[end])
        }
    }

    private fun perpendicularDistance(
        point: LocationData,
        lineStart: LocationData,
        lineEnd: LocationData
    ): Double {
        // Calculate perpendicular distance from point to line
        val x0 = point.latitude
        val y0 = point.longitude
        val x1 = lineStart.latitude
        val y1 = lineStart.longitude
        val x2 = lineEnd.latitude
        val y2 = lineEnd.longitude

        val numerator = abs((y2 - y1) * x0 - (x2 - x1) * y0 + x2 * y1 - y2 * x1)
        val denominator = sqrt((y2 - y1).pow(2) + (x2 - x1).pow(2))

        if (denominator == 0.0) {
            return calculateDistance(point, lineStart)
        }

        // Convert to meters (rough approximation)
        val degreesToMeters = 111319.9 // meters per degree at equator
        return (numerator / denominator) * degreesToMeters
    }
}

// MARK: - Spatial Grid for Efficient Proximity Queries

class SpatialGrid {
    companion object {
        private const val CELL_SIZE = 50.0 // 50 meters per cell
        private const val METERS_PER_DEGREE = 111319.9
    }

    private val grid = mutableMapOf<GridCell, MutableList<Int>>()

    data class GridCell(val x: Int, val y: Int)

    fun insert(point: LocationData, index: Int) {
        val cell = getCell(point)
        grid.getOrPut(cell) { mutableListOf() }.add(index)
    }

    fun findNearbyPoints(point: LocationData, radius: Double): List<Int> {
        val centerCell = getCell(point)
        val nearbyIndices = mutableListOf<Int>()

        // Check 3x3 grid of cells
        for (dx in -1..1) {
            for (dy in -1..1) {
                val cell = GridCell(centerCell.x + dx, centerCell.y + dy)
                grid[cell]?.let { nearbyIndices.addAll(it) }
            }
        }

        return nearbyIndices
    }

    private fun getCell(point: LocationData): GridCell {
        // Convert lat/lon to approximate meters (rough calculation)
        val x = (point.latitude * METERS_PER_DEGREE / CELL_SIZE).toInt()
        val y = (point.longitude * METERS_PER_DEGREE / CELL_SIZE).toInt()
        return GridCell(x, y)
    }
}
